### PR TITLE
Derivatives trading logics

### PR DESCRIPTION
**Pull request by JoylimJY**
- refactor pool component
- refactor option/power component
- add general trading functions for future-option-power (@/service/derivatives.js)

### CODE REVIEW

File: src/components/OptionComponent/index.js [hunk 1]
```diff
@@ -28,45 +30,19 @@ const OptionComponent = props => {
...
     selectedToken,
     symbol,
-    onTrade,
   } = props
 
-  const { account, library, farmSetting: { INITIAL_ALLOWED_SLIPPAGE } } = useConstantLoader(props);
-  const { chainId } = useChainId();
-  const connectWalletByLocalStorage = useConnectWallet();
-  const { active } = useWeb3React();
+  const { account, farmSetting: { INITIAL_ALLOWED_SLIPPAGE } } = useConstantLoader(props)
```
Comment (ykchong45) [inline src/components/OptionComponent/index.js:35]: `constantLoader` is no longer in use.

**Comment by ykchong45** [comment]
## Issues

1. `OptionComponent` not responsive to user input / percentage change

2. Pool page needs refactoring, lots of logics for button, text and error can be simplified. Use of `Alp` should be checked thoroughly, and `addLiquidity` function is not yet called in the UI.

3. We need a default provider when user is not connected to the wallet (mostly useful for read functions)

File: src/acy-dex-futures/utils/index.js [hunk 2]
```diff
@@ -17,7 +17,7 @@ import { getContract } from './Addresses';
 import { useLocalStorage } from "react-use";
 import { constantInstance, useConstantLoader } from '@/constants';
 import { format as formatDateFn } from "date-fns";
-import { BURN_FEE_BASIS_POINTS, MINT_FEE_BASIS_POINTS } from './_Helpers';
+import { MINT_FEE_BASIS_POINTS } from './_Helpers';
```
Comment (ykchong45) [inline src/acy-dex-futures/utils/index.js:20]: In the end we'd like to remove all files starting with `_`, so if you find any imports from such files, please move the code to other files.

File: src/components/AcyPoolComponent/index.js [hunk 3]
```diff
@@ -1,104 +1,482 @@
...
+} from '@/acy-dex-futures/utils';
+import PerpetualTabs from '@/components/PerpetualComponent/components/PerpetualTabs';
+import BuyInputSection from '@/pages/BuyGlp/components/BuyInputSection';
+import glp40Icon from '@/pages/BuyGlp/components/ic_glp_40.svg'
+import AcyPerpetualButton from '@/components/PerpetualComponent/components/AcyPerpetualButton';
+import ERC20 from '@/acy-dex-futures/abis/ERC20.json';
+import Token from '@/acy-dex-futures/abis/Token.json';
+import Router from '@/abis/future-option-power/Router.json';
+import Reader from '@/abis/future-option-power/Reader.json';
+import Alp from '@/abis/future-option-power/Alp.json'
```
Comment (ykchong45) [inline src/components/AcyPoolComponent/index.js:34]: In current smart contract, we use [`MintableToken`](https://github.com/ACY-Labs/acy-futures-options-powers/blob/develop_liquidityV2/contracts/pool/PoolImplementation.sol#L382) as template to create LP token. Thus we should use [`MintableToken`](https://github.com/ACY-Labs/acy-futures-options-powers/blob/develop_liquidityV2/contracts/token/MintableToken.sol) as LP token's ABI.

File: src/components/AcyPoolComponent/index.js [hunk 4]
```diff
@@ -1,104 +1,482 @@
...
-  // const glpAddress = perpetuals.getContract("GLP")
-  const readerAddress = getContract(chainId, "Reader");
-  const vaultAddress = getContract(chainId, "Vault");
-  const usdgAddress = getContract(chainId, "USDG");
-  const nativeTokenAddress = getContract(chainId, "NATIVE_TOKEN");
-  const routerAddress = getContract(chainId, "Router");
-  const glpManagerAddress = getContract(chainId, "GlpManager");
-  const glpAddress = getContract(chainId, "GLP");
-  const rewardRouterAddress = getContract(chainId, "RewardRouter");
+  const { account } = useConstantLoader(props)
```
Comment (ykchong45) [inline src/components/AcyPoolComponent/index.js]: We don't use `useContantLoader()`, and `account` can be retrieved from `useWeb3React()`, as in line 43 below.

File: src/components/OptionComponent/index.js [hunk 5]
```diff
@@ -133,6 +134,7 @@ const OptionComponent = props => {
 
   useEffect(() => {
     setShowDescription(false)
+    setMarginToken(tokens[1])
   }, [chainId, mode])
```
Comment (ykchong45) [inline src/components/OptionComponent/index.js]: [Check dependency array](https://devtrium.com/posts/dependency-arrays#what-should-be-put-in-the-dependency-array)

File: src/services/derivatives.js [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  token,
+  tokenAmount,
+) {
+  const contract = new ethers.Contract(routerAddress, Router.abi, library.getSigner())
+
+  let method = "addMargin"
+  let params = [
+    token.address,  //token address
+    ethers.utils.parseUnits(tokenAmount, token.decimals),  //amount
+    token.symbol,  //token symbol
```
Comment (ykchong45) [inline src/services/derivatives.js]: This should be the trading symbol name, i.e. "BTCUSD" or "BTCUSD-60000-C", not the symbol of deposit token.

File: src/components/AcyPoolComponent/index.js (hunk 3 above)
Comment (ykchong45) [reply src/components/AcyPoolComponent/index.js:34]: Joy update: `Alp.json` is actually `MintableToken`'s abi format.

File: src/components/OptionComponent/index.js [hunk 7]
```diff
@@ -139,9 +139,9 @@ const OptionComponent = props => {
       return
     }
     if (mode == 'Buy') {
-      trade(chainId, library, poolAddress, IPool, account, symbol, selectedTokenAmount, symbolMarkPrice)
+      trade(chainId, library, poolAddress, IPool, account, symbol, selectedTokenAmount, symbolMarkPrice?.mul(bigNumberify(10050)).div(bigNumberify(10000)))
     } else {
-      trade(chainId, library, poolAddress, IPool, account, symbol, selectedTokenAmount.mul(bigNumberify(-1)), symbolMarkPrice)
+      trade(chainId, library, poolAddress, IPool, account, symbol, selectedTokenAmount.mul(bigNumberify(-1)), symbolMarkPrice?.mul(bigNumberify(9950)).div(bigNumberify(10000)))
     }
```
Comment (ykchong45) [inline src/components/OptionComponent/index.js:148]: trade still might fail due to the incorrect pricing calculation in the smart contract, especially for option symbols. @ykchong45 will check.

File: src/components/AcyPoolComponent/index.js [hunk 8]
```diff
@@ -1,104 +1,502 @@
...
+  const selectedTokenBalance = tokenInfo?.find(item => item.token?.toLowerCase() == selectedToken.address?.toLowerCase())?.balance
+  const amountUsd = selectedTokenPrice?.mul(selectedTokenAmount)
 
-  const { data: glpBalance, mutate: updateGlpBalance } = useSWR([chainId, glpAddress, "balanceOf", account || PLACEHOLDER_ACCOUNT], {
-    fetcher: fetcher(library, Glp),
-  })
+  let feePercentageText = formatAmount(feeBasisPoints, 2, 2, true, "-")
+  if (feeBasisPoints !== undefined && feeBasisPoints.toString().length > 0) {
+    feePercentageText += "%"
+  }
```
Comment (ykchong45) [inline src/components/AcyPoolComponent/index.js:133]: I'm not sure if a react state would be a good use case for this case 🤔  What's your thought?

File: src/components/AcyPoolComponent/index.js [hunk 9]
```diff
@@ -1,104 +1,502 @@
...
-        infoTokens={infoTokens}
-        glpPrice={glpPrice}
-        glpBalance={glpBalance}
-        glpBalanceUsd={glpBalanceUsd}
-        glpSupply={glpSupply}
-        glpSupplyUsd={glpSupplyUsd}
-      />
+      <div className="GlpSwap">
+        <div className={styles.BuySellSelector}>
+          <PerpetualTabs
```
Comment (ykchong45) [inline src/components/AcyPoolComponent/index.js]: Rename the component would be nice. 😄

File: src/components/OptionComponent/index.js [hunk 10]
```diff
@@ -63,113 +107,44 @@ const OptionComponent = props => {
...
-    if (needApproval) {
-      return `Approve ${selectedToken.symbol}`;
-    }
     return mode == 'Buy' ? 'Buy / Long' : 'Sell / Short'
   }
 
+  useEffect(() => {
+    setShowDescription(false)
+    setMarginToken(tokens[1])
+  }, [chainId, tokens])
```
Comment (ykchong45) [inline src/components/OptionComponent/index.js]: 🤔  Is the `chainId` necessary here? `tokens` is a state that changes with `chainId` already.

File: src/components/OptionComponent/index.js (hunk 7 above)
Comment (ykchong45) [reply src/components/OptionComponent/index.js:148]: 👍  It turns out that having 0.5% slippage when trading (either long or short) will give us successful trade() call.

**Review by ykchong45 — CHANGES_REQUESTED** [review]

File: src/acy-dex-futures/utils/index.js (hunk 2 above)
Comment (ykchong45) [reply src/acy-dex-futures/utils/index.js:20]: to YK: Should have this in the future refactor list, not an urgent task to be done at the moment.

File: src/components/AcyPoolComponent/index.js (hunk 8 above)
Comment (JoylimJY) [reply src/components/AcyPoolComponent/index.js:133]: feeBasisPoints is a react state. feePercentageText is just a text for display, so no need to use react state?

**Review by ykchong45 — APPROVED** [review]

**Pull request merged**